vpc/service/assign_addresses_v3.go

```diff
@@ -98,27 +102,56 @@ type branchENI struct {
 	idx           int
 }
 
+func subnetCacheKey(az, accountID string, subnetIDs []string) string {
+	sort.Strings(subnetIDs)
+	return fmt.Sprintf("az:%s accountID:%s subnetIDs:%s", az, accountID, strings.Join(subnetIDs, ","))
+}
+
 func (vpcService *vpcService) getSubnet(ctx context.Context, az, accountID string, subnetIDs []string) (*subnet, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	ctx, span := trace.StartSpan(ctx, "getSubnet")
-	defer span.End()
+	// NOTE[jigish] caching using the full set of subnetIDs will cause the same subnet to be returned for all hits
+	cacheKey := subnetCacheKey(az, accountID, subnetIDs)
+	r, err, _ := vpcService.getSubnetLock.Do(cacheKey, func() (interface{}, error) {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		ctx, span := trace.StartSpan(ctx, "getSubnet")
+		defer span.End()
+
+		item := vpcService.getSubnetCache.Get(cacheKey)
+		var staleSubnet *subnet
+		if item != nil {
+			span.AddAttributes(trace.BoolAttribute("cached", true))
+			span.AddAttributes(trace.BoolAttribute("expired", item.Expired()))
+			span.AddAttributes(trace.StringAttribute("expires", item.Expires().String()))
+
+			val := item.Value().(*subnet)
+			if !item.Expired() {
+				return val, nil
+			} else if item.Expires().Add(subnetStaleValidTime).After(time.Now()) {
```

Comment:
Can you double check this?

If we get an item and cache it at T=0 (minutes)
It expires at T=1
It is stale up to T=11 (expires.add)

If it is now T=15, should we execute line 130? Am I understanding "After" correctly? Oh, no, yea, because 11.After(15) is false.

Wait, I lost it. I'm not sure why the word doesn't work so well in my brain.  I got it again, whew. ! before(now) doesn't read any better to me, idk.

👍

Reply:
`time.Now().Before(...)` would read better maybe?

Reply:
Oh yea :) 
All good though, I'm not that worried.